src/core/lib/event_engine/windows/windows_engine.cc

```diff
@@ -140,15 +178,171 @@ bool WindowsEventEngine::IsWorkerThread() {
+                nullptr, 0, nullptr, info->overlapped());
+  // It wouldn't be unusual to get a success immediately. But we'll still get an
+  // IOCP notification, so let's ignore it.
+  if (!success) {
+    int last_error = WSAGetLastError();
+    if (last_error != ERROR_IO_PENDING) {
+      auto status = GRPC_WSA_ERROR(WSAGetLastError(), "ConnectEx");
+      Run([on_connect = std::move(on_connect), status]() mutable {
+        on_connect(status);
+      });
+      watched_socket->MaybeShutdown(status);
+      return invalid_connection_handle;
+    }
+  }
+  GPR_ASSERT(watched_socket != nullptr);
+  auto connection_state = std::make_shared<ConnectionState>();
+  grpc_core::MutexLock lock(&connection_state->mu);
+  connection_state->address = address;
+  connection_state->socket = std::move(watched_socket);
+  connection_state->on_connected_user_callback = std::move(on_connect);
+  connection_state->allocator = std::move(memory_allocator);
+  connection_state->on_connected =
+      SelfDeletingClosure::Create([this, connection_state]() mutable {
+        OnConnectCompleted(std::move(connection_state));
+      });
+  {
+    grpc_core::MutexLock conn_lock(&connection_mu_);
+    connection_state->connection_handle =
+        ConnectionHandle{reinterpret_cast<intptr_t>(connection_state.get()),
+                         aba_token_.fetch_add(1)};
+    known_connection_handles_.insert(connection_state->connection_handle);
+  }
+  connection_state->timer_handle =
+      RunAfter(timeout, [this, connection_state]() {
+        if (CancelConnect(connection_state->connection_handle)) {
+          // delete the other ref to the connection state
+          delete connection_state->on_connected;
+          connection_state->on_connected_user_callback(
+              absl::DeadlineExceededError("Connection timed out"));
+        }
+        // else: The connection attempt could not be canceled. We can assume the
+        // connection callback will be called.
+      });
+  connection_state->socket->NotifyOnWrite(connection_state->on_connected);
+  return connection_state->connection_handle;
+}
+
+bool WindowsEventEngine::CancelConnect(EventEngine::ConnectionHandle handle) {
+  if (TaskHandleComparator<ConnectionHandle>::Eq()(handle,
+                                                   invalid_connection_handle)) {
+    GRPC_EVENT_ENGINE_TRACE("%s",
+                            "Attempted to cancel an invalid connection handle");
+    return false;
+  }
+  grpc_core::MutexLock lock(&connection_mu_);
+  if (!known_connection_handles_.contains(handle)) {
+    GRPC_EVENT_ENGINE_TRACE(
+        "Unknown connection handle: %s",
+        HandleToString<EventEngine::ConnectionHandle>(handle).c_str());
+    return false;
+  }
+  auto* connection_state = reinterpret_cast<ConnectionState*>(handle.keys[0]);
+  connection_state->socket->MaybeShutdown(
+      absl::CancelledError("CancelConnect"));
+  known_connection_handles_.erase(handle);
```

Comment:
I think if CancelConnect succeeds, the timer started in line 309 also needs to be cancelled. Otherwise the timer callback will execute after deadline expires and it may not delete connection_state->on_connected.

Reply:
Good catch, fixed.